test/tangent_types/abstract_zero.jl

```diff
@@ -160,3 +160,156 @@
+            contents::Number
+            MyPartiallyDefinedStruct(x) = new(x)
+        end
+        dy = zero_tangent(MyPartiallyDefinedStruct(1.5))
+        @test iszero(dy.intro)
+        @test iszero(dy.contents)
+        @test (dy.contents = 2.0) == 2.0  # should be assignable
+
+        mutable struct MyPartiallyDefinedStructWithAnys
+            intro::Float64
+            contents::Any
+            MyPartiallyDefinedStructWithAnys(x) = new(x)
+        end
+        dy = zero_tangent(MyPartiallyDefinedStructWithAnys(1.5))
+        @test iszero(dy.intro)
+        @test iszero(dy.contents)
+        @test dy.contents === ZeroTangent()  # we just don't know anything about this data
+        @test (dy.contents = 2.0) == 2.0  # should be assignable
+        @test (dy.contents = [2.0, 4.0]) == [2.0, 4.0]  # should be assignable to different values
+
+        mutable struct MyStructWithNonConcreteFields
+            x::Any
+            y::Union{Float64,Vector{Float64}}
+            z::AbstractVector
+        end
+        d = zero_tangent(MyStructWithNonConcreteFields(1.0, 2.0, [3.0]))
+        @test iszero(d.x)
+        d.x = Tangent{Base.RefValue{Float64}}(; x=1.5)
+        @test d.x == Tangent{Base.RefValue{Float64}}(; x=1.5)  #should be assignable
+        d.x = 2.4
+        @test d.x == 2.4  #should be assignable
+        @test iszero(d.y)
+        d.y = 2.4
+        @test d.y == 2.4  #should be assignable
+        d.y = [2.4]
+        @test d.y == [2.4]  #should be assignable
+        @test iszero(d.z)
+        d.z = [1.0, 2.0]
+        @test d.z == [1.0, 2.0]
+        d.z = @view [2.0, 3.0, 4.0][1:2]
+        @test d.z == [2.0, 3.0]
+        @test d.z isa SubArray
+    end
+
+    @testset "cyclic references" begin
+        mutable struct Link
+            data::Float64
+            next::Link
+            Link(data) = new(data)
+        end
+
+        lk = Link(1.5)
+        lk.next = lk
+
+        d = zero_tangent(lk)
+        @test d.data == 0.0
+        @test d.next === d
+
+        # The following two cases are broken
+        # We hope they are not too significant, because in general if you AD step by step they should work
+        # (as should the one above so maybe we should get rid of this extra complex logic)
+        # It's only a problem if you first do the multistep build then `zero_tangent` rather than `zero_tangent` at the constructor.
+
+        # Idea: check if `!isbitstype` only if so do we need to worry about caching etc
+        struct CarryingArray
+            x::Vector
+        end
+        ca = CarryingArray(Any[1.5])
+        push!(ca.x, ca)
+        @test_broken d_ca = zero_tangent(ca)
+        @test_broken d_ca[1] == 0.0
+        @test_broken d_ca[2] === _ca
+
+        # Idea: check if typeof(xs) <: eltype(xs), if so need to cache it before computing
+        xs = Any[1.5]
+        push!(xs, xs)
+        @test_broken d_xs = zero_tangent(xs)
+        @test_broken d_xs[1] == 0.0
+        @test_broken d_xs[2] == d_xs
```

Comment:
These are truly evil. Good test cases

Reply:
yep, I am not going to fix them in this PR but I will leave them here marked broken

Reply:
Perfect